Implement delist strategy

msperetto opened the pull request:
This PR will do:

- Adjust table from Database to have now operation types to be identified when running code. Could be "trading", "delist", "arbitrage", etc...
- Start using strategies from database, instead of manually adding it to code;
- Implement function to monitor and get binance announcements related to completed delisted coins;
- Implement function to identify if the coin is being traded in futures market;
- Implement separation of "trading" execution to "delist" execution;
- Checking balance and all initial setup;
- Open and closing delist operation;
- Notifying and Saving those operations;

msperetto committed:
Add html code to show the main dashboard for backtest

msperetto committed:
Add plotly (js library) to plot graphs to dashboard

msperetto committed:
Add domain for backtests results

msperetto committed:
Add initial dao to get all backtest results

msperetto committed:
Add API endpoint to get backtest results

msperetto committed:
Fix endpoint name to get backtest results

msperetto committed:
Fix api get backtest function to make sure it's getting data correctly

msperetto committed:
Refactor html and js to show more relevant data

msperetto committed:
Fix response type to be a list

Also add debugging messages in log

msperetto committed:
Fix row variable name to avoid UnboundLocalError with module rows

msperetto committed:
Fix backtest float fields with nan or inf values to be JSON compliant

msperetto committed:
Refactor FASTapi to allow cross-origin requests from JS dashboard

msperetto committed:
Add debug message on Loading data on Javascript

msperetto committed:
Refactor adjusting size of bubbles in bubble chart

ltoledozt committed:
Refactor moving scripts related to Backtest DB to be in the correct folder

Since we have different directories for Backtest and Prod, I think it makes more sense keeping the files and folders different.
So, this commit:
- Drops the script related to backtest that was added in Prod file.
- Created some SQL script files that might be useful for Backtest. Probably in the future.
- Write the Alter Table logic related to Backtest in the correct place.

ltoledozt committed:
Fix DB table created with the wrong name

I could just rename the table, but since we are kind a starting a new dashboard for backtest, I think it's fine starting the table from scratch.
So, also included the new `Side` column to be added as part of the table creation.
Note: I did some small adjustments in the SQL script, like:
	• created_at: to use 'with' time zone (keep consistent with the other datetime fields of this table - start_time/end_time)
	• best_indicators_combination: to use TEXT instead of varying(6000)

I also did some adjustments in other places that were using the wrong name.

ltoledozt committed:
Drop unnecessary SQL script

I believe that file was created when we started working with Backtest.
Since now we have a better SQL scripts folder structure, I believe we can drop it.

msperetto committed:
Refactor values from backtest to get from the view with positive results

msperetto committed:
Add side column to backtest (sanitize and dao)

msperetto committed:
add column score to view and api

msperetto committed:
Add new charts do dashboard

msperetto committed:
Add ability to easily see the date range when analyzing backtest results

Improve readability of backtest logs by setting human-readable period label.
Note: In theory we could use the existing 'start_time', but there's a logic related to `longest_indicator` and `extended_start_time` that makes harder to figure out.

msperetto committed:
Add operation_type column for exchange_config and strategy tables

Also adding all necessary dependencies

msperetto committed:
Add getting Delist announcements and check if coin in futures

ltoledozt commented on main/common/delist.py:
```diff
@@ -0,0 +1,54 @@
+
+            if link and "Binance Will Delist" in title:
+                # Get the next sibling element, which contain the date of the announcement
+                announcement_date = announcement.find_next_sibling().get_text(strip=True)
```
Qual data é essa exatamente? 

<img width="377" height="65" alt="image" src="https://github.com/user-attachments/assets/cae4b2ba-17a8-4e60-b9c3-30254a42a375" />

Nesse caso acima, entendo que pegaria a data `2025-09-02`, confere?

ltoledozt commented on main/common/delist.py:
```diff
@@ -0,0 +1,54 @@
+                # Get the next sibling element, which contain the date of the announcement
+                announcement_date = announcement.find_next_sibling().get_text(strip=True)
+                today = datetime.now().strftime("%Y-%m-%d")
+                if announcement_date == today:
```
To pensando num cenário alternativo. Será que existe o risco da Binance inserir essa notificação perto do final do dia? Tipo: "2025-09-02 23:59:00"

Minha preocupação é por uma questão de minutos ou mesmo de timezone, o Marlin (🤖🐟) não conseguir identificar essa atualização no site.. e aí não conseguir entrar na operação!

Algumas ideias:
- Talvez ser mais permissivo e incluir mais datas pra executar? (ex.: today, dia seguinte da notificação? dia seguinte dentro de um intervalo de horas?)
- Talvez termos uma tabela interna de controle de Delist? Por exemplo, quando vier uma atualização de Delist, o Marlin inclui essa Delist numa tabela de controle e com status se já foi capturado ou não. Exemplo: Tabela Delist (id, data, title, status). Esse status seria 0 (não foi processado) ou 1 (já foi processado). Esse conceito de status poderia permitir ser mais "permissivo" e não checar apenas o "today". Entende?

(Não sei se isso aqui é muito overkilling). O ponto é mais diminuir o risco de aparecer uma notificação de Delist nova e o Marlin não conseguir processar por motivos de data ou algo parecido!

ltoledozt commented on main/common/delist.py:
```diff
@@ -0,0 +1,54 @@
+        # geting all <a> tags
+        return soup.find_all("a")
+
+    def extract_tickers_from_title(title):
```
Seria interessante ter Unit Test pra esse método hein. 
Talvez até pra alguns outros dessa classe!

ltoledozt commented on main/common/delist.py:
```diff
@@ -0,0 +1,54 @@
+    def get_all_futures_symbols():
+        return Binance().get_all_futures_symbols()
+
+    def check_ticker_in_futures(ticker):
```
O que esse metodo retorna exatamente? 
Um `summary comment` nesses métodos seria util!

ltoledozt commented on main/common/delist.py:
```diff
@@ -0,0 +1,54 @@
+            tickers = re.findall(r"\b[A-Z]{2,}\b", tickers_text)  # Extract tickers
+            return tickers
+
+    def get_all_futures_symbols():
```
Me parece que esse é um método usado apenas dentro dessa classe. Nesse caso, sugiro nomeá-lo como  `_get_all_futures_symbols`. Aí fica mais evidente que é um método privado.

msperetto committed:
Refactor comparing new annoucement with DB instead of today() date

- Add table, dao and domain for delist_announcements;
- If announcement is new, already add it to DB;
- Add missing "selfs";
- adding _ to identify private methods;

msperetto committed:
Add summary comments for important methods

msperetto committed:
Fix get_all_symbols() function typo name

msperetto committed:
Create Unit Test to check Delist().check_ticker_in_futures

msperetto committed:
Refactor way to get binance delist announcements

First was web scrapping;

Now utilizing the internal Binance API endpoint (same that is being utilized in the Binance delist announcement page)

msperetto committed:
Add unit test to test if getting binance announcements correctly

msperetto committed:
Add a backup url to get delist announcements

msperetto committed:
Add a Listing annoucements method

Return a list of newest listing announcements

msperetto committed:
Refactor delisting unit tests to be grouped in only one class

Make it easier to maintain, grow and test

msperetto committed:
Add get new listings unit test

msperetto committed:
Refactor get_binance_announcements to be private method

msperetto committed:
Refactor Delist to check coins already delisted, instead of date

Added all necessary methods and adjustments

msperetto committed:
Refactor method name to be easier to Louis understandment

msperetto committed:
Refactor - fine tunning parameters

Now the call will just return the last 2 announcements

ltoledozt commented on main/prod/binance.py:
```diff
@@ -393,3 +396,80 @@ def cancel_all_open_orders(self, symbol, b_id, b_sk):
+        else:
+            # try backup url
+            response = requests.get(self.BINANCE_ANNOUNCEMENTS_URL_2, params=params_backup, headers=headers)
+            if response.status_code != 200:
```
Reparei que na chamada de cima, é checado `response.status_code == 200` e nessa `response.status_code != 200`. Uma sugestão: ter esse tipo de checagem de forma consistente. Isso deixaria o código mais legível e fácil de dar manutenção.

Nesse caso, ficaria:

```
        response = requests.get(self.BINANCE_ANNOUNCEMENTS_URL, params=params, headers=headers)
        if response.status_code == 200:
            data = response.json()
            announcements = data.get("data", {}).get("articles", [])
            return announcements
        else:
            # try backup url
            response = requests.get(self.BINANCE_ANNOUNCEMENTS_URL_2, params=params_backup, headers=headers)
            if response.status_code == 200:
                data = response.json()
                announcements = data.get("data", {}).get("catalogs", {})[0].get("articles", [])
                return announcements
            else:
                logger.error(f'Error fetching delist announcements: {response.status_code}')
                return None
```

ltoledozt commented on main/prod/binance.py:
```diff
@@ -393,3 +396,80 @@ def cancel_all_open_orders(self, symbol, b_id, b_sk):
+            return announcements
+
+
+    def get_list_announcements(self):
```
Ao invés de definir `params`, `params_backup` e `headers` (tudo isso pra diferentes chamadas), ficaria melhor definir encapsular as chamadas da API em métodos privados.  Dessa forma respeitariamos o principio SRP (Single Responsibility Principle).

Sugestão:

```
def get_delist_announcements(self):
        """
        Fetch and parse Binance delisting announcements.
        """
        announcements = self._fetch_primary()
        if announcements is None:
            announcements = self._fetch_backup()

        return announcements
```

```
    def _fetch_primary(self):
        params = {
            "pageNo": 1,
            "pageSize": 2,
            "catalogId": 161
        }
        try:
            response = requests.get(self.BINANCE_ANNOUNCEMENTS_URL,
                                    params=params,
                                    headers=self._headers,
                                    timeout=10)
            response.raise_for_status()
            data = response.json()
            return data.get("data", {}).get("articles", [])
        except Exception as e:
            logger.warning(f"Primary fetch failed: {e}")
            return None
```

```
def _fetch_backup(self):
        params = {
            "type": 1,
            "pageNo": 1,
            "pageSize": 2,
            "catalogId": 161
        }
        try:
            response = requests.get(self.BINANCE_ANNOUNCEMENTS_URL_2,
                                    params=params,
                                    headers=self._headers,
                                    timeout=10)
            response.raise_for_status()
            data = response.json()
            return data.get("data", {}).get("catalogs", [{}])[0].get("articles", [])
        except Exception as e:
            logger.error(f"Backup fetch failed: {e}")
            return None

```

Como o headers é comum para ambas chamadas, poderia defini-la num lugar comum (ex.: no `init` da classe).

```
    def __init__(self):
        self._headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                          "AppleWebKit/537.36 (KHTML, like Gecko) "
                          "Chrome/58.0.3029.110 Safari/537.3",
            "Accept": "application/json, text/plain, */*",
            "Accept-Language": "en-US,en;q=0.5",
            "Connection": "keep-alive"
        }
```

**Obs.:** Na prática, quando usamos `response.raise_for_status()`, não precisamos checar manualmente if `response.status_code == 200`, porque esse método já lança uma exceção (HTTPError) se o status não estiver no range 200–299.
Ou seja, o try/except com raise_for_status() cobre todos os casos de erro (400, 404, 500 etc.).

ltoledozt replied:
Reparei que os métodos `get_delist_announcements` e `get_list_announcements`. 
Acredito que esse refactoring que sugeri vai até facilitar deixar os métodos principais reutizando os 2 novos métodos privados. 
Confere?

msperetto replied:
nem conhecia esse raise_for_status() .... muito espertinho esse chatgpt

msperetto committed:
Refactor way to get new announcements to follow SRP

added 2 fetching private methods and a common _header private attribute to smoothly run principal and backup requests

msperetto committed:
Fix delist dao adding missing DEV_ENV_CON

msperetto committed:
temp add binance new listing tracking

This commit is only to confirm the viability of getting announcements from binance;
The hypothesis is that if the time from the announcement is close enough from the time our bot fetch the information, it is a viable strategy;

ltoledozt commented on main/common/dao/list_announcement_dao.py:
@msperetto Fioo, ta certo colocar as alterações do arquivo `list_announcement_dao.py`? 
Tenho a impressão que na verdade esse arquivo deveria estar no outro commit, nao nesse!
Confere? 
(Nao sei se isso se aplica aos demais arquivos desse commit)

msperetto committed:
Refactor way to instantiate Strategies when starting the bot

We used no hard code a list of strategies objects;

Now we get the enabled strategy name from DB, inform the type of the strategy ("trading", "delist", "list", etc...) and instantiate it

msperetto committed:
Move delist file into prod folder

msperetto committed:
Drop unused Beautifulsoup import

msperetto committed:
Add dao function to get trades by operation type

Operation types can be, for example:
"trading"; "delist"; "list"

msperetto committed:
Add simple trade sell class, to be used to open delist trades

msperetto committed:
Add class to be run every hour to update all balances from sub accounts

msperetto committed:
Add way to separate sub accounts when running different modules;

Enable possibility to run several different subaccounts to different marlin methods;

msperetto committed:
Add new loggers

msperetto committed:
Add new tests to check account balances are being updated correctly

msperetto committed:
Add SQL trigger to automatically update date field on accnt_balnce table

msperetto committed:
Add simple Delist strategy

Drop unnecessary indicator and fix trend_class name

Add sma indicator for trend

msperetto committed:
Add Delist functionality and all necessary dependencies

ltoledozt commented on main/prod/update_account_balances.py:
```diff
@@ -0,0 +1,57 @@
+                    existing_balance = get_account_balance(account_id)
+                    
+                    if existing_balance:
+                        update_account_balance(account_id, account_balance, margin_ratio=2)  # Assuming margin_ratio is 2 for simplicity
```
I would suggest setting the `margin_ratio` using a `const` in this class.

ltoledozt commented on main/prod/update_account_balances.py:
```diff
@@ -0,0 +1,57 @@
+        self.exchange_sessions = exchange_sessions
+        self.binance = Binance()
+
+    def update_account_balances():
```
Missing `self` parameter.

@msperetto Are you facing some issues with your Visual Code? In my environment, it clears complaints about some compilation errors in this file.

<img width="1302" height="833" alt="image" src="https://github.com/user-attachments/assets/83ee8013-4bb8-4544-abe6-2c5fd11b60cd" />

ltoledozt commented on main/prod/delist.py:
In my VS, I can see some compilation issues. Can you in your environment?

<img width="1082" height="550" alt="image" src="https://github.com/user-attachments/assets/a23f017a-0305-4e21-8648-4722116defaa" />

<img width="1182" height="407" alt="image" src="https://github.com/user-attachments/assets/5f5fc8ce-9b8e-4d3b-b0e8-35bc9b33eb0c" />

<img width="1242" height="800" alt="image" src="https://github.com/user-attachments/assets/a5079b44-ffbc-4ffd-888d-c488c3a83462" />

And some other issues...

ltoledozt commented on main/prod/delist.py:
```diff
@@ -1,15 +1,26 @@
+        self.strategy = strategy # We'll have only one strategy for delist
+        self.exchange_session = exchange_session
+        self.available_balance = self._update_available_balance()
+        self.setup = None # It's a 
```
Missing the rest of the comment.

msperetto committed:
Fix imports and it's utilizations

msperetto committed:
Fix the utilization of the private method including self

msperetto committed:
Add new app to run delist as a docker service

Also add service configuration in docker compose file

msperetto committed:
Add new app to run update_account_balances as a docker service

Add service configuration in docker compose file
Fix small errors in update_balances main file

msperetto committed:
Refactor _get_strategies_by_type method from app to be DRY

It's going to be utilized by other "app" files, like app_delist.py

msperetto committed:
Drop not used code

Code was intended to run delist and trading_bot paralleled

msperetto committed:
Refactor way of getting secrets to work with multiple accounts

msperetto committed:
Drop unused imports

msperetto committed:
Update pandas_ta library version

Adjusting pandas_ta version to be compatible with python 3.11

msperetto committed:
Update dockerfile to improve security measures

ltoledozt commented on main/prod/app_delist.py:
@msperetto tava pensando aqui.. Essa pasta `prod` ta ficando meio extensa.
Será que não ficaria melhor, se organizassemos em subpastas?
Tipo, "delist", "trading", etc.
Arquivos que não se encaixam num assunto especifico, aí OK deixarmos debaixo de `prod` mesmo. Mas arquivos q é exclusivo pra um assunto, talvez vale a pena organizar!
O que vc acha?

msperetto committed:
Refactor Tech analyses by using ta-lib instead of pandas_ta

Pandas TA was presenting conflicts with python 3.11, so we need to substitute it by a more robust library

msperetto committed:
Refactor ta library again - from talib to python ta

TA-Lib also gave lots of conflicts with C and the library wheel-build

msperetto committed:
Add mock test to test delist operations;

This commit does:
    - create a new table to create fake announcements for testing purposes;
    - create api endpoints to interact with table mentioned above;
    - Create dao and model necessary to run the test;
    - Create a delist method to capture it from the api;
    - Separate test and prod environments;
    - Do all other necessary configurations

Refactor delist to guarantee that tests will be run exactly like prod

msperetto committed:
Refactor dockerfile to run psycopg correctly

msperetto committed:
Add telegram env variable to docker compose services

msperetto committed:
Fix typo in ACCOUNT_ID_DELIST import

msperetto committed:
Drop unnecessary duplicated line to update balances in db

msperetto committed:
Temp chancing NEGOTIATION_ENV variable to TEST delist

msperetto committed:
Add margin_ratio regular binance value for simplicity

msperetto committed:
Add coin to delist announcement dao

msperetto committed:
Add more logging for debug

Add debug logging

msperetto committed:
Fix passing trend param

msperetto committed:
Fix prefix issue when merging datasets

msperetto committed:
Fix passing pairs instead of coins to methods

Methods affected:
 - get_pairs_precision();
 - get_pairs_price_precision();

Also dropped non necessary loggers

msperetto committed:
Fix way to get "side" to mock delist

msperetto committed:
Drop unused name for mock test

msperetto committed:
Add instructions to copilot agent

msperetto committed:
Refactor test delist to mark not delisting announcement as processed

This enable the code to stop processing repeated annoucement

ltoledozt commented on main/prod/delist.py:
```diff
@@ -0,0 +1,413 @@
+            try:
+                # In TEST mode, simulate the position instead of opening it
+                if NEGOCIATION_ENV == Environment_Type.TEST:
+                    # Calculate quantity based on order value and current price
```
The code below could be encapsulated in a new private method. It makes the code cleaner, more readable, and eaiser to maintain. 
Something like:

```
  # In TEST mode, simulate the position instead of opening it
  if NEGOCIATION_ENV == Environment_Type.TEST:
      self._log_simulated_position(manager, binance, pair, coin);
      continue
```

msperetto committed:
Drop debug loggings

msperetto committed:
Fix import and self errors

Errors my IDE was not showing in the past

msperetto committed:
Add integration test for delist

The idea is to simulate new announcement inputs, check if it's running properly, generate a summary of all tests and cleanup everything in the end;

Everything is documented in *.md files
